internal/terraform/node_resource_plan_instance_query.go

```diff
@@ -0,0 +1,101 @@
+
+package terraform
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/internal/providers"
+	"github.com/hashicorp/terraform/internal/states"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+func (n *NodePlannableResourceInstance) listResourceExecute(ctx EvalContext) (diags tfdiags.Diagnostics) {
+	log.Printf("[TRACE] NodePlannableResourceInstance: listing resources for %s", n.Addr)
+	config := n.Config
+	addr := n.ResourceInstanceAddr()
+	provider, providerSchema, err := getProvider(ctx, n.ResolvedProvider)
+	diags = diags.Append(err)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// validate self ref
+	diags = diags.Append(validateSelfRef(addr.Resource, config.Config, providerSchema))
+	if diags.HasErrors() {
+		return diags
+	}
+
+	keyData := EvalDataForInstanceKey(addr.Resource.Key, nil)
+	if config.ForEach != nil {
+		forEach, _, _ := evaluateForEachExpression(config.ForEach, ctx, false)
+		keyData = EvalDataForInstanceKey(addr.Resource.Key, forEach)
+	}
+
+	// evaluate the list config block
+	var configDiags tfdiags.Diagnostics
+	configVal, _, configDiags := ctx.EvaluateBlock(config.Config, n.Schema.Body, nil, keyData)
+	diags = diags.Append(configDiags)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// Unmark before sending to provider
+	unmarkedConfigVal, _ := configVal.UnmarkDeepWithPaths()
+	configKnown := configVal.IsWhollyKnown()
+	if !configKnown {
+		diags = diags.Append(fmt.Errorf("config is not known"))
+		return diags
+	}
+
+	log.Printf("[TRACE] NodePlannableResourceInstance: Re-validating config for %s", n.Addr)
+	validateResp := provider.ValidateListResourceConfig(
+		providers.ValidateListResourceConfigRequest{
+			TypeName: n.Config.Type,
+			Config:   unmarkedConfigVal,
+		},
+	)
+	diags = diags.Append(validateResp.Diagnostics.InConfigBody(config.Config, n.Addr.String()))
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// If we get down here then our configuration is complete and we're ready
+	// to actually call the provider to list the data.
+	resp, err := provider.ListResource(providers.ListResourceRequest{
+		TypeName: n.Config.Type,
+		Config:   unmarkedConfigVal,
+	})
+	if err != nil {
+		return diags.Append(fmt.Errorf("failed to list %s: %s", n.Addr, err))
+	}
+
+	resources := make([]cty.Value, 0)
+	identities := make([]cty.Value, 0)
+
+	for evt := range resp {
+		if evt.Diagnostics.HasErrors() {
+			return diags.Append(evt.Diagnostics.InConfigBody(config.Config, n.Addr.String()))
+		}
+		resources = append(resources, evt.ResourceObject)
+		identities = append(identities, evt.Identity)
+	}
+
+	var vals, ids cty.Value
+	if len(resources) > 0 {
+		vals = cty.ListVal(resources)
+		ids = cty.ListVal(identities)
+	} else {
+		vals = cty.ListValEmpty(cty.Object(map[string]cty.Type{}))
+		ids = cty.ListValEmpty(cty.Object(map[string]cty.Type{}))
+	}
+
+	ctx.State().SetListResourceInstance(n.Addr, &states.ResourceInstanceObject{
+		Value:    vals,
```

Comment:
This is going to make evaluating a list resource difficult, because it's a list rather than an object. This is why we were thinking about making the list resource type match the schema, and holding the results as a special attribute of that object.

Reply:
While we store this as a list/tuple, we did compose in the evaluation context as an object
https://github.com/hashicorp/terraform/blob/15ca83851d0a2ad549324e623391ceb48b71987d/internal/terraform/evaluate.go#L821-L826

I mostly used it this way, so that we can reuse the state object. Perhaps it would be better to introduce a new object entirely.

Reply:
oh interesting 🤔 While there's not really any value in referencing list resource config attributes, that is how the rest of Terraform works so we may want to keep it consistent. One example I can think of is if the list resource has a `data` attribute:
```
list "provider_thing" "resources" {
  data = ["some", "data"]
}

output "provider_things" {
  value = list.provider_thing.resources.data
}
```

A terraform user would expect that output reference to contain `["some", "data"]`. That means we either need to reserve keywords from the schema, or move config into a reserved block 🤔

Reply:
If I have to use the changes object instead, I will keep this in mind.

Reply:
Oh good point. Perhaps we need to retain that `data`, and potentially `identity`

Reply:
The alternative from a consistency point of view is to explicitly name the config block:
```
list "provider_thing" "resources" {
  config {
    data = ["some", "data"]
  }
}
```
We discussed how that is different from normal resources too, but maybe because list resource always have these "computed attributes" containing the results they just have to be different